processing/src/main/java/org/apache/druid/query/UnionQueryRunner.java

```diff
@@ -41,30 +42,27 @@ public UnionQueryRunner(
   public Sequence<T> run(final QueryPlus<T> queryPlus, final ResponseContext responseContext)
   {
     Query<T> query = queryPlus.getQuery();
-    DataSource dataSource = query.getDataSource();
-    if (dataSource instanceof UnionDataSource) {
+
+    final DataSourceAnalysis analysis = DataSourceAnalysis.forDataSource(query.getDataSource());
+
+    if (analysis.isConcreteTableBased() && analysis.getBaseTableDataSources().get().size() != 1) {
```

Comment:
can we use more explicit condition such as `analysis.getBaseDataSource() instanceof UnionDataSource` than this `analysis.getBaseTableDataSources().get().size() != 1`?  May be the assumption around number of tables will not be valid in future.

Reply:
Hmm, I was trying to avoid using `instanceof` checks outside of the DataSourceAnalysis class, and instead stick to cleanly defined interfaces as much as possible. So, I designed dataSourceAnalysis.getBaseTableDataSources() to encapsulate this logic (it has the `instanceof` check you're talking about).

Do you think it'd be better to instead move the logic to UnionQueryRunner?

Reply:
In that case, It would also be fine to have a static method `isUnionDataSource()` within `DataSourceAnalysis` class itself that can be used here as well as in `dataSourceAnalysis.getBaseTableDataSources()`

Reply:
Hmm, in thinking about your comment I realized there is a bug here. The code doesn't work right if you have a UnionDataSource of a single TableDataSource. I'll fix it and add a test for this, and while doing so, I'll consider your comments about how to structure the code.

Thanks for the comments!

Reply:
I pushed up a new commit that changes the approach a bit; let me know what you think. The `getBaseTableDataSources()` concept is gone and replaced by two separate methods, `getBaseTableDataSource()` and `getBaseUnionDataSource()`.

Reply:
Looks good now.

Reply:
Great!